akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala

```diff
@@ -66,44 +131,77 @@ private[remote] final class TopHeavyHitters[T >: Null](val max: Int)(implicit cl
+        // This is not guaranteed to be the one we are searching for, yet (hash values may collide).
+        // From this position we can invoke the more costly search which checks actual object equalities.
+        // With this two step search we avoid equality checks completely for many non-colliding entries.
+        val actualIdx = findItemIdx(candidateIndex, hashCode, item)
+
+        // usually O(1), worst case O(n) if we need to scan due to hash conflicts
+        if (actualIdx == -1) {
+          // So we don't have this entry so far (only a colliding one, it was a false positive from findHashIdx).
+          insertKnownNewHeavy(hashCode, item, count) // O(1 + log n), we simply replace the current lowest heavy hitter
           true
-        case potentialIndexGuess ⇒
-          // the found index could be one of many which hash to the same value (we're using open-addressing),
-          // so it is only used as hint for the replace call. If the value matches, we're good, if not we need to search from here onwards.
-          val actualIdx = findItemIdx(potentialIndexGuess, hashCode, item)
-
-          if (actualIdx == -1) {
-            insertKnownNewHeavy(hashCode, item, count) // O(1 + log n), we simply replace the current lowest heavy hitter
-            true
-          } else replaceExistingHeavyHitter(actualIdx, hashCode, item, count) // usually O(1), worst case O(n) if we need to scan due to hash conflicts
+        } else {
+          // The entry exists, let's update it.
+          updateExistingHeavyHitter(actualIdx, hashCode, item, count)
+          // not a "new" heavy hitter, since we only replaced it (so it was signaled as new once before)
+          false
+        }
       }
+
     }
 
+  /**
+   * Checks the lowest weight entry in this structure and returns true if the given count is larger than that. In
+   * other words this checks if a new entry can be added as it is larger than the known least weight.
+   */
   def isHeavy(count: Long): Boolean =
     count > lowestHitterWeight
 
+  /**
+   * Finds the index of an element in the hashtable (or returns -1 if it is not found). It is usually a good idea
+   * to find the first eligible index with findHashIdx before invoking this method since that finds the first
+   * eligible candidate (one with an equal hash value) faster than this method (since this checks actual object
+   * equality).
+   */
   private def findItemIdx(searchFromIndex: Int, hashCode: HashCodeVal, o: T): Int = {
     @tailrec def loop(index: Int, start: Int, hashCodeVal: HashCodeVal, o: T): Int = {
+      // Scanned the whole table, returned to the start => element not in table
       if (index == start) -1
-      else if (hashCodeVal.get == hashes(index)) {
+      else if (hashCodeVal.get == hashes(index)) { // First check on hashcode to avoid costly equality
         val item: T = items(index)
-        if (item == o) {
+        if (Objects.equals(item, o)) {
```

Comment:
must have been some code moved from java I guess.
It's a null-safe way to `.equals()` in java.